augur/filter.py

```diff
@@ -884,91 +884,72 @@ def get_groups_for_subsampling(strains, metadata, group_by=None):
     [{'strain': 'strain1', 'filter': 'skip_group_by_with_ambiguous_month', 'kwargs': ''}]
 
     """
-    if group_by:
-        groups = group_by
-    else:
-        groups = ("_dummy",)
-
+    metadata = metadata.loc[strains]
     group_by_strain = {}
     skipped_strains = []
-    for strain in strains:
-        skip_strain = False
-        group = []
-        m = metadata.loc[strain].to_dict()
-        # collect group specifiers
-        for c in groups:
-            if c == "_dummy":
-                group.append(c)
-            elif c in m:
-                group.append(m[c])
-            elif c in ['month', 'year'] and 'date' in m:
-                try:
-                    year = int(m["date"].split('-')[0])
-                except:
+
+    if not group_by or group_by == ('_dummy',):
+        group_by_strain = {strain: ('_dummy',) for strain in strains}
+        return group_by_strain, skipped_strains
+
+    groups = group_by
+    groups_set = set(groups)
+
+    date_requested = ('year' in groups_set or 'month' in groups_set)
+
+    # If we could not find any requested categories, we cannot complete subsampling.
+    if 'date' not in metadata and groups_set <= {'year', 'month'}:
+        raise FilterException(f"The specified group-by categories ({groups}) were not found. No sequences-per-group sampling will be done. Note that using 'year' or 'year month' requires a column called 'date'.")
+    if not groups_set & (set(metadata.columns) | {'year', 'month'}):
+        raise FilterException(f"The specified group-by categories ({groups}) were not found. No sequences-per-group sampling will be done.")
+
+    if date_requested:
+        if 'date' not in metadata:
+            # set year/month/day = unknown
+            print(f"WARNING: A 'date' column could not be found to group-by year or month.", file=sys.stderr)
+            print(f"Filtering by group may behave differently than expected!", file=sys.stderr)
+            df_dates = pd.DataFrame({'year': 'unknown', 'month': 'unknown'}, index=metadata.index)
+            metadata = pd.concat([metadata, df_dates], axis=1)
+        else:
+            # replace date with year/month/day as nullable ints
+            date_cols = ['year', 'month', 'day']
+            df_dates = (metadata['date'].str.split('-', n=2, expand=True)
+                                            .set_axis(date_cols, axis=1))
+            for col in date_cols:
+                df_dates[col] = pd.to_numeric(df_dates[col], errors='coerce').astype(pd.Int64Dtype())
```

Comment:
Hmm, this is broken with incomplete dates (e.g. just `2020`). I'll write some tests and fix.

Reply:
False alarm – this is fine and there was another problem. Nevertheless, added a test here: https://github.com/nextstrain/augur/pull/794/commits/5596dd4bbcaab721b33b27206f6c1ef7e75a2bce